oauth2_client/models.py

```diff
@@ -1,54 +1,250 @@
-    created_at = models.DateTimeField(auto_now_add=timezone.now)
+    extra_settings = JSONField(default=dict, blank=True)
 
     def __unicode__(self):
         """
         Returns the application name to show in the admin.
         """
         return self.name
 
+    def __str__(self):
+        return "<Applicaiton id:%s, name:%s>" % (self.id, self.name)
+
+    def validate_jwt_grant(self):
+        """
+        If JWT token bearer grant is used, user email has to be specified either by a FK or using
+        extra_settings['USER'] field.
+        """
+        from django.core.exceptions import ValidationError
+        if self.authorization_grant_type == Application.GRANT_JWT_BEARER and \
+                ('USER' not in self.extra_settings and not get_value_or_default(self, None, 'user', 'email')):
+            msg = (
+                "grant_type={} requires user's email to be specified either in extra_settings or by a FK field `user`"
+                .format(Application.GRANT_JWT_BEARER)
+            )
+            raise ValidationError(msg)
+
+    def clean(self):
+        self.validate_jwt_grant()
+
+
+class MobileApplicationAdapter:
+    """
+    An adapter class to expose Application for the mobile_auth package in LTI repo.
+    """
+
+    def __init__(self, app):
+        self.app = app
+
+    @property
+    def name(self):
+        return self.app.name
+
+    @property
+    def key(self):
+        return self.app.client_id
+
+    @property
+    def secret(self):
+        return self.app.client_secret
+
+
+class TokenExpiryState:
+    VALID = 'valid'         # token not expired (including TIMEOUT_SECONDS margin)
+    UNKNOWN = 'unknown'     # no expiration info on token
+    EXPIRED = 'exired'      # token expired (including TIMEOUT_SECONDS margin)
+    ABSENT = 'absent'       # we have no token
+
 
 class AccessToken(models.Model):
     """
     Access token for the service provider.
     """
-    application = models.ForeignKey(Application, on_delete=models.CASCADE)
-    # store in JSON because OAuth2Session and BackendApplicationClient require it in this format
-    token = JSONField()
-    created_at = models.DateTimeField(auto_now_add=timezone.now, db_index=True)
+
+    """
+    Used to prevent the token expiration when the request is processed on
+    the server if the token is expiring in seconds or less at the client
+    request time.
+    """
+    TIMEOUT_SECONDS = float(60)
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name="%(app_label)s_%(class)s"
+    )
+    token = models.CharField(
+        max_length=255,
+        unique=True
+    )
+    raw_token = JSONField(
+        default=dict,
+        blank=True
+    )
+    token_type = models.TextField()
+    application = models.ForeignKey(
+        Application,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
```

Comment:
why `null=True` and `blank=True` are added?

Reply:
Because that class started its life as a copy of an equivalent from `oauth2_provider` and it seems I didn't clean that up thoroughly enough.
Fixed